nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ControllerServiceResource.java

```diff
@@ -680,6 +681,116 @@ public Response updateControllerService(
+        } catch (final IllegalArgumentException iae) {
+            // ignore
+        }
+
+        // ensure an action has been specified
+        if (requestControllerServiceState == null) {
+            throw new IllegalArgumentException("Must specify the updated state. To update the referencing Controller Services the "
+                    + "state should be DISABLED.");
+        }
+
+        // ensure the controller service state is not DISABLING
+        if (!ControllerServiceState.DISABLED.equals(requestControllerServiceState) ) {
+            throw new IllegalArgumentException("Cannot set the referencing services to ENABLING or DISABLING");
+        }
+
+        final Revision requestRevision = getRevision(requestControllerServiceEntity, id);
+        return withWriteLock(
+                serviceFacade,
+                serviceFacade.getControllerService(id, true),
+                requestRevision,
+                lookup -> {
+                    // authorize the service
+                    final ComponentAuthorizable authorizable = lookup.getControllerService(requestControllerServiceDTO.getId());
+                    authorizable.getAuthorizable().authorize(authorizer, RequestAction.WRITE, NiFiUserUtils.getNiFiUser());
+
+                    // authorize the current and new process groups
+                    final ProcessGroupAuthorizable authorizableProcessGroupNew = lookup.getProcessGroup(requestControllerServiceEntity.getParentGroupId());
+                    authorizableProcessGroupNew.getAuthorizable().authorize(authorizer, RequestAction.WRITE, NiFiUserUtils.getNiFiUser());
+
+                    final ProcessGroupAuthorizable authorizableProcessGroupOld = lookup.getProcessGroup(requestControllerServiceDTO.getParentGroupId());
+                    authorizableProcessGroupOld.getAuthorizable().authorize(authorizer, RequestAction.WRITE, NiFiUserUtils.getNiFiUser());
+
+                    // authorize any referenced services
+                    AuthorizeControllerServiceReference.authorizeControllerServiceReferences(requestControllerServiceDTO.getProperties(), authorizable, authorizer, lookup);
+                    AuthorizeParameterReference.authorizeParameterReferences(requestControllerServiceDTO.getProperties(), authorizer, authorizable.getParameterContext(),
+                            NiFiUserUtils.getNiFiUser());
+
+                    // authorize referencing components
+                    requestControllerServiceDTO.getReferencingComponents().forEach(e -> {
+                        final Authorizable controllerService = lookup.getControllerServiceReferencingComponent(requestControllerServiceDTO.getId(), e.getId());
+                        OperationAuthorizable.authorizeOperation(controllerService, authorizer, NiFiUserUtils.getNiFiUser());
+                    });
+                },
+                null,
```

Comment:
Verification step should go here, to ensure that any components that are referencing the Controller Service will still be able to reference the Controller Service once it has been moved. Otherwise, it should throw an IllegalStateException.

Reply:
The move process now fails and shows an error message when there is a scope conflict with a referencing component. References are no longer removed.  Validation is indeed done here to check for conflicts.